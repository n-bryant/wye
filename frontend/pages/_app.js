import React from "react";
import App from "next/app";

import Page from "../src/components/Page";
import { ApolloProvider } from "react-apollo";
import withData from "../lib/withData";

import CssBaseline from "@material-ui/core/CssBaseline";

import NextNProgress from "../src/components/NextNProgress";

// set up a context provider
const AppContext = React.createContext({});
export const AppContextProvider = AppContext.Provider;
export const AppContextConsumer = AppContext.Consumer;

// reducer action and content options
export const ACTIONS = {
  SET_USERS: "setUsers",
  SET_CONTENT: "setContent"
};
export const CONTENT_OPTIONS = {
  WELCOME: "WELCOME",
  FORM: "FORM",
  RECOMMENDATIONS: "RECOMMENDATIONS"
};

// initial state values
export const INITIAL_STATE = {
  content: CONTENT_OPTIONS.WELCOME,
  users: []
};

/**
 * reducer for updating state values
 * @param {Object} state
 * @param {Object} action
 * @returns {Object}
 */
export const reducer = (state, action) => {
  switch (action.type) {
    case ACTIONS.SET_USERS:
      return {
        ...state,
        users: action.value
      };
    case ACTIONS.SET_CONTENT:
      return {
        ...state,
        content: action.value
      };
    default:
      throw new Error();
  }
};

export class WyeApp extends App {
  componentDidMount() {
    // Remove the server-side injected CSS.
    const jssStyles = document.querySelector("#jss-server-side");
    if (jssStyles) {
      jssStyles.parentElement.removeChild(jssStyles);
    }
  }

  static async getInitialProps({ Component, ctx }) {
    let pageProps = {};
    if (Component.getInitialProps) {
      pageProps = await Component.getInitialProps(ctx);
    }
    // expose query to the user
    pageProps.query = ctx.query;
    return { pageProps };
  }

  render() {
    return <WyeWithState {...this.props} />;
  }
}

// app with state
export const WyeWithState = props => {
  const { Component, apollo, pageProps } = props;
  const [state, dispatch] = React.useReducer(reducer, INITIAL_STATE);

  return (
    <React.Fragment>
      <CssBaseline />
      <NextNProgress />
      <ApolloProvider client={apollo}>
        <AppContextProvider
          value={{
            state,
            dispatch
          }}
        >
          <Page>
            <Component {...pageProps} />
          </Page>
        </AppContextProvider>
      </ApolloProvider>
    </React.Fragment>
  );
};

// app with apollo client
export const WyeWithData = withData(WyeApp);

export default WyeWithData;
