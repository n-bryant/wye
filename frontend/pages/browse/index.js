import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../lib/classNamePrefix";
import { withWidth } from "@material-ui/core/";
import { withStyles } from "@material-ui/core/styles";

import Container from "@material-ui/core/Container";
import Typography from "@material-ui/core/Typography";
import BackgroundProvider from "../../src/components/BackgroundProvider";
import styles from "./index.styles";

/**
 * renders the Browse page
 */
export const BrowsePage = props => {
  const classnames = BrowsePage.classnames(props);

  //                              top right of page ->     advanced search
  // Heading - Browse Multiplayer Games
  // Subheading - Browse the most popular and discounted multiplayer games

  // trending - most popular by recent playtime
  // - Cards: Contain appropriate capsule image, name, and price details
  //    - on hover: shows name, user rating, dev/pub info and
  // capsuleLg - 1:4

  // publishers - games by most popular publishers
  // - Cards: 2 pages of 4 cards; page paddles and page pills
  //  - Header image linking to most popular title by publisher
  //  - Publisher name linking to /publisher/[name]

  // big sales - top 4 - rated by discount and user rating
  // - Cards: 2 pages of 3 cards; page paddles and page pills
  //  - libraryCapsule image on top
  //  - name / sale details on bottom
  //  - on hover: - name, user rating, user details if available (wanted by)

  // buttons to go to filter for under $5 / $10

  // Browse by Genre
  // - grid of genre buttons

  return (
    <BackgroundProvider>
      <div className={classnames.root()}>
        <Container
          className={classnames.element("container")}
          maxWidth="lg"
          disableGutters
        >
          <Container className={classnames.element("main")} maxWidth={false}>
            <Typography className={classnames.element("heading")} variant="h1">
              Featured Games
            </Typography>
            <Typography
              className={classnames.element("subHeading")}
              variant="h2"
            >
              Browse the most popular and discounted multiplayer games
            </Typography>
          </Container>
        </Container>
      </div>
    </BackgroundProvider>
  );
};
BrowsePage.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}BrowsePage`
);
BrowsePage.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    container: PropTypes.string,
    main: PropTypes.string,
    heading: PropTypes.string,
    subHeading: PropTypes.string
  })
};
BrowsePage.defaultProps = {
  classes: {}
};

export default withWidth()(withStyles(styles)(BrowsePage));
