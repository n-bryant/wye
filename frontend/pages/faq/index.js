import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../lib/classNamePrefix";
import { withStyles } from "@material-ui/core/styles";

import FaqAccordion from "../../src/components/FaqAccordion";
import styles from "./index.styles";

/**
 * renders an FAQ page
 */
export const FaqPage = props => {
  const classnames = FaqPage.classnames(props);

  return (
    <div className={classnames.root()}>
      <FaqAccordion />
    </div>
  );
};
FaqPage.classnames = createClassNameHelper(`${JSS_CLASS_NAME_PREFIX}FaqPage`);
FaqPage.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string
  })
};
FaqPage.defaultProps = {
  classes: {}
};

export default withStyles(styles)(FaqPage);
