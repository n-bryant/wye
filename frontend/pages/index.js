import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../lib/classNamePrefix";

import Icon from "@mdi/react";
import { mdiGithubCircle } from "@mdi/js";
import { AppBar, Container, Toolbar, Typography } from "@material-ui/core";

import Link from "next/link";
import WelcomeEmptyState from "../src/components/WelcomeEmptyState";
import RecommendationsForm from "../src/components/RecommendationsForm/RecommendationsForm";
import IndexPageBackground from "../src/components/ScrollingBackground/IndexPageBackground";
import styles from "./index.styles";

// set up a context provider
const AppContext = React.createContext({});
export const AppContextProvider = AppContext.Provider;
export const AppContextConsumer = AppContext.Consumer;

// initial state values
export const INITIAL_STATE = {
  showForm: false,
  users: []
};
/**
 * reducer for updating state values
 * @param {Object} state
 * @param {Object} action
 * @returns {Object}
 */
export const reducer = (state, action) => {
  switch (action.type) {
    case "setShowForm":
      return {
        ...state,
        showForm: action.value
      };
    case "setUsers":
      return {
        ...state,
        users: action.value
      };
    default:
      throw new Error();
  }
};

/**
 * renders the index page's content,
 * dynamically displaying either the welcome content or the recommendations form
 */
export const Index = props => {
  const classnames = Index.classnames(props);
  const [state, dispatch] = React.useReducer(reducer, INITIAL_STATE);

  return (
    <AppContextProvider
      value={{
        state,
        dispatch
      }}
    >
      <div className={classnames.root()}>
        <Container
          className={classnames.element("container")}
          maxWidth={false}
          disableGutters
        >
          <AppBar position="static" className={classnames.element("appBar")}>
            <Toolbar className={classnames.element("toolBar")}>
              <Typography className={classnames.element("title")} variant="h1">
                Wye
              </Typography>
              <div className={classnames.element("linksContainer")}>
                <Link href="/faq">
                  <a className={classnames.element("link")}>faq</a>
                </Link>
                <Link href="https://github.com/n-bryant/wye" prefetch={false}>
                  <a className={classnames.element("link")} target="_blank">
                    <Icon
                      className={classnames.element("githubIcon")}
                      path={mdiGithubCircle}
                    />
                  </a>
                </Link>
              </div>
            </Toolbar>
          </AppBar>
          <div className={classnames.element("main")}>
            <div className={classnames.element("contentContainer")}>
              <div className={classnames.element("content")}>
                {state.showForm ? (
                  <RecommendationsForm className={classnames.element("form")} />
                ) : (
                  <WelcomeEmptyState />
                )}
              </div>
            </div>
          </div>
        </Container>
        <IndexPageBackground />
      </div>
    </AppContextProvider>
  );
};
Index.classnames = createClassNameHelper(`${JSS_CLASS_NAME_PREFIX}Index`);
Index.propTypes = {
  // styles applied
  classes: PropTypes.shape({
    root: PropTypes.string,
    container: PropTypes.string,
    appBar: PropTypes.string,
    toolBar: PropTypes.string,
    title: PropTypes.string,
    linksContainer: PropTypes.string,
    link: PropTypes.string,
    githubIcon: PropTypes.string,
    form: PropTypes.string,
    main: PropTypes.string,
    contentContainer: PropTypes.string,
    content: PropTypes.string
  })
};
Index.defaultProps = {
  classes: {}
};

export default withStyles(styles)(Index);
