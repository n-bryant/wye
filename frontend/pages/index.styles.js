const styles = theme => ({
  root: {
    height: "100%",
    position: "relative",
    overflow: "hidden"
  },
  container: {
    backgroundColor: "rgba(34, 139, 230,.25)",
    height: "100%",
    zIndex: "1"
  },
  main: {
    height: "100%",
    overflow: "auto"
  },
  contentContainer: {
    height: "100%",
    width: "100%",
    maxWidth: "1500px",
    minHeight: "530px",
    margin: "0 auto",
    padding: "3rem",
    [theme.breakpoints.down("xs")]: {
      padding: "0"
    }
  },
  content: {
    height: "100%",
    width: "100%",
    background: "rgba(0,0,0,0.8)",
    backdropFilter: "blur(6px)",
    border: "1px solid hsla(0,0%,100%,0.2)",
    overflow: "auto"
  }
});
export default styles;
