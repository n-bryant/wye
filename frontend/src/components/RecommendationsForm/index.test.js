import { Formik } from "formik";
import {
  RecommendationsFormWithContext,
  RecommendationsFormWithFormik,
  RECOMMENDATIONS_FORM_INITIAL_VALUES,
  RecommendationsForm
} from "./index";
import { ACTIONS, CONTENT_OPTIONS } from "../../../pages";

describe("RecommendationsFormWithContext", () => {
  it("should provide a dispatch function from context and pass any additional props to its rendered child", () => {
    const props = {
      foo: "bar"
    };

    const renderCallbackArgs = {
      dispatch: jest.fn()
    };

    const wrapper = shallow(<RecommendationsFormWithContext {...props} />);
    expect(wrapper.prop("children")(renderCallbackArgs).props.dispatch).toBe(
      renderCallbackArgs.dispatch
    );
    expect(wrapper.prop("children")(renderCallbackArgs).props.foo).toBe(
      props.foo
    );
  });
});

describe("RecommendationsFormWithFormik", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<RecommendationsFormWithFormik />);
  });

  it("should render a Formik and pass it initialValues from props or a default initialValues prop matching RECOMMENDATIONS_FORM_INITIAL_VALUES", () => {
    expect(wrapper.find(Formik).prop("initialValues")).toMatchObject(
      RECOMMENDATIONS_FORM_INITIAL_VALUES
    );

    const newProps = {
      initialValues: {
        users: ["foo", "bar"]
      }
    };
    wrapper.setProps(newProps);
    wrapper.update();
    expect(wrapper.find(Formik).prop("initialValues")).toMatchObject(
      newProps.initialValues
    );
  });

  it("should pass a dispatch prop to its rendered child", () => {
    const props = {
      dispatch: jest.fn()
    };
    wrapper.setProps(props);
    expect(wrapper.prop("children")().props.dispatch).toBe(props.dispatch);
  });
});

describe("RecommendationsForm", () => {
  const props = {
    classes: {
      root: "root",
      closeButton: "closeButton",
      formInstructions: "formInstructions",
      title: "title",
      userFieldContainer: "userFieldContainer",
      submitButton: "submitButton"
    },
    dispatch: jest.fn(),
    values: {
      users: []
    }
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<RecommendationsForm {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should call the dispatch prop to show the welcome content when the close button is clicked", () => {
    wrapper
      .findWhere(n => n.hasClass(props.classes.closeButton))
      .simulate("click");
    expect(props.dispatch).toHaveBeenCalledWith({
      type: ACTIONS.SET_CONTENT,
      value: CONTENT_OPTIONS.WELCOME
    });
  });

  it("should set the disabled value of the ActionButton based on whether the users value of the values prop is defined and has length", () => {
    // without users length
    expect(
      wrapper
        .findWhere(n => n.hasClass(props.classes.submitButton))
        .prop("disabled")
    ).toBeTruthy();

    // undefined users
    wrapper.setProps({
      values: {}
    });
    wrapper.update();
    expect(
      wrapper
        .findWhere(n => n.hasClass(props.classes.submitButton))
        .prop("disabled")
    ).toBeTruthy();

    // users defined and with length
    wrapper.setProps({
      values: {
        users: ["foo"]
      }
    });
    wrapper.update();
    expect(
      wrapper
        .findWhere(n => n.hasClass(props.classes.submitButton))
        .prop("disabled")
    ).toBeFalsy();
  });
});
