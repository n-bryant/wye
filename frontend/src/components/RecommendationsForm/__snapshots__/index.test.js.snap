// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecommendationsForm should render successfully 1`] = `
<div
  className="wye-RecommendationsForm root"
>
  <Icon
    className="wye-RecommendationsForm__closeButton closeButton"
    color="currentColor"
    horizontal={false}
    onClick={[Function]}
    path="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
    rotate={0}
    size={null}
    spin={false}
    vertical={false}
  />
  <div
    className="wye-RecommendationsForm__formInstructions formInstructions"
  >
    <WithStyles(ForwardRef(Typography))
      className="wye-RecommendationsForm__title title"
      variant="h1"
    >
      Getting Your Recommendations:
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      Wye will compare the libraries of each provided SteamID to generate your recommendations. By default, Wye will recommend the top rated multiplayer game owned by the majority of the entered SteamIDs, but additional filters will become available after the initial search. Use the field below to add up to six SteamIDs, then click "Go!".
    </WithStyles(ForwardRef(Typography))>
  </div>
  <div
    className="wye-RecommendationsForm__userFieldContainer userFieldContainer"
  >
    <WithStyles(UsersField)
      setUsersValue={[Function]}
      usersValue={Array []}
    />
  </div>
  <WithStyles(ActionButton)
    className="wye-RecommendationsForm__submitButton submitButton"
    disabled={true}
    onClick={[Function]}
  />
</div>
`;
